smart-contracts/contracts/Unlock.sol

```diff
@@ -126,7 +148,55 @@ contract Unlock is
     initializer()
   {
     // We must manually initialize Ownable
-    Ownable.__initializeOwnable(_unlockOwner);
+    UnlockOwnable.__initializeOwnable(_unlockOwner);
+  }
+
+  function initializeProxyAdmin() public {
+    require(proxyAdminAddress == address(0), "ProxyAdmin already deployed");
+    _deployProxyAdmin();
+  }
+
+  /**
+  * @dev Deploy the ProxyAdmin contract that will manage lock templates upgrades
+  * This deploys an instance of ProxyAdmin used by PublicLock transparent proxies.
+  */
+  function _deployProxyAdmin() private returns(address) {
+    proxyAdmin = new ProxyAdmin();
+    proxyAdminAddress = address(proxyAdmin);
+    return address(proxyAdmin);
+  }
+
+  /**
+  * @dev Helper to get the version number of a template from his address
+  */
+  function publicLockVersions(address _impl) external view returns(uint16) {
+    return _publicLockVersions[_impl];
+  }
+
+  /**
+  * @dev Helper to get the address of a template based on its version number
+  */
+  function publicLockImpls(uint16 _version) external view returns(address) {
+    return _publicLockImpls[_version];
+  }
+
+  /**
+  * @dev Registers a new PublicLock template immplementation
+  * The template is identified by a version number
+  * Once registered, the template can be used to upgrade an existing Lock
+  */
+  function addLockTemplate(address impl, uint16 version) public onlyOwner {
+    require(impl != address(0), "impl address can not be 0x");
+    require(version != 0, "version can not be 0");
+    require(_publicLockVersions[impl] == 0, "address already used by another version");
+    require(_publicLockImpls[version] == address(0), "version already assigned");
+
+    _publicLockVersions[impl] = version;
+    _publicLockImpls[version] = impl;
+    if (publicLockLatestVersion < version) publicLockLatestVersion = version;
+
+    // TODO: init template to prevent possible attack
```

Comment:
👍  but that means these params have to be passed as `data` to this function?

Reply:
it should be doable on the FE

Reply:
And would not be that different from how multisg wallets and/pr Governor contracts actually work (submitting a tx that includes a tx to be sent if approved by other members)

Reply:
yep agree. The main issue is that we will have to trigger the upgrade off-chain

Reply:
wdym?